Opt-in wide-layout attachment style using Shadow DOM tree

squelart opened the pull request:
#### d706e84b583afba61b95a266f34c5059df7e1d54
<pre>
Opt-in wide-layout attachment style using Shadow DOM tree
<a href="https://bugs.webkit.org/show_bug.cgi?id=251830">https://bugs.webkit.org/show_bug.cgi?id=251830</a>
rdar://105103464

Reviewed by Aditya Keerthi.

Note that the modern attachment element uses an inner legacy attachment just to display the correct thumbnail or icon.

* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::HTMLAttachmentElement):
(WebCore::HTMLAttachmentElement::ensureModernShadowRoot):
(WebCore::HTMLAttachmentElement::didAddUserAgentShadowRoot):
(WebCore::HTMLAttachmentElement::ensureModernShadowTree):
(WebCore::HTMLAttachmentElement::createElementRenderer): Deleted.
(WebCore::HTMLAttachmentElement::getAttachmentIdentifier): Deleted.
(WebCore::HTMLAttachmentElement::copyNonAttributePropertiesFromElement): Deleted.
(WebCore::HTMLAttachmentElement::archiveResourceURL): Deleted.
(WebCore::HTMLAttachmentElement::file const): Deleted.
(WebCore::HTMLAttachmentElement::blobURL const): Deleted.
(WebCore::HTMLAttachmentElement::setFile): Deleted.
(WebCore::HTMLAttachmentElement::insertedIntoAncestor): Deleted.
(WebCore::HTMLAttachmentElement::removedFromAncestor): Deleted.
(WebCore::HTMLAttachmentElement::ensureUniqueIdentifier): Deleted.
(WebCore::HTMLAttachmentElement::setUniqueIdentifier): Deleted.
(WebCore::HTMLAttachmentElement::enclosingImageElement const): Deleted.
(WebCore::HTMLAttachmentElement::parseAttribute): Deleted.
(WebCore::HTMLAttachmentElement::attachmentTitle const): Deleted.
(WebCore::HTMLAttachmentElement::attachmentTitleForDisplay const): Deleted.
(WebCore::HTMLAttachmentElement::attachmentType const): Deleted.
(WebCore::HTMLAttachmentElement::attachmentPath const): Deleted.
(WebCore::HTMLAttachmentElement::updateAttributes): Deleted.
(WebCore::mimeTypeIsSuitableForInlineImageAttachment): Deleted.
(WebCore::HTMLAttachmentElement::updateEnclosingImageWithData): Deleted.
(WebCore::HTMLAttachmentElement::updateThumbnail): Deleted.
(WebCore::HTMLAttachmentElement::updateIcon): Deleted.
(WebCore::HTMLAttachmentElement::requestIconWithSize const): Deleted.
(WebCore::HTMLAttachmentElement::childShouldCreateRenderer const): Deleted.
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/rendering/AttachmentLayout.mm:
(WebCore::AttachmentLayout::layOutSubtitle):
(WebCore::AttachmentLayout::AttachmentLayout):

Canonical link: <a href="https://commits.webkit.org/260296@main">https://commits.webkit.org/260296@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2440ed660b7dd90e4823f522bb4299e7b6cf1164

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107464 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16526 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40367 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116627 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116048 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111362 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17951 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7768 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99628 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113225 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13552 "Found 5 new test failures: imported/w3c/web-platform-tests/css/css-counter-styles/devanagari/css3-counter-styles-119.html, imported/w3c/web-platform-tests/css/css-counter-styles/devanagari/css3-counter-styles-120.html, imported/w3c/web-platform-tests/css/css-counter-styles/devanagari/css3-counter-styles-121.html, imported/w3c/web-platform-tests/html/cross-origin-opener-policy/header-parsing-failures.https.html?12-last, webgl/webgl-ext-norm16-texture-texsubimage-nocrash.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96733 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41179 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95457 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28359 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/82950 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/96801 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9532 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29712 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/96223 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/7531 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10189 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6611 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30807 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15688 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49293 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/105071 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7123 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11749 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26059 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -624,6 +626,19 @@ AttachmentElementEnabled:
+  status: embedder
+  humanReadableName: "Attachment wide-layout styling"
+  humanReadableDescription: "Use horizontal wide-layout attachment style, implies Attachment Element"
+  webcoreBinding: DeprecatedGlobalSettings
```
We should avoid new uses of `DeprecatedGlobalSettings`.

You can remove this line, and add a:

```
WebCore:
      default: false
```

Then, access the new setting via `Settings` (which is accessible through `Document`).

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -70,14 +75,122 @@ HTMLAttachmentElement::~HTMLAttachmentElement() = default;
+    if (m_elementWithTitle)
+        return;
+
+    constexpr auto styleSheet = R"CSS(
```
I saw your comment that said you'd move this to a separate file in a followup, but I don't see why this can't be done now. If you still want to do it separately, please file a bug and leave a `FIXME` here.

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -70,14 +75,122 @@ HTMLAttachmentElement::~HTMLAttachmentElement() = default;
+attachment::-webkit-attachment-container {
+    display: grid;
+    max-width: 500px;
+    background-color: lightgray;
```
This is fine for an initial patch, but please file a bug and leave a `FIXME` here to use semantic colors (adapt to dark mode / accessibility settings)

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -70,14 +75,122 @@ HTMLAttachmentElement::~HTMLAttachmentElement() = default;
+    -webkit-user-select: none;
+}
+attachment::selection {
+    background-color: blue;
```
Same comment about semantic colors.

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -70,14 +75,122 @@ HTMLAttachmentElement::~HTMLAttachmentElement() = default;
+    grid-row: 3;
+    grid-column: 2;
+    font-size: x-small;
+    color: rgb(82, 145, 214);
```
Same comment about semantic colors.

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -70,14 +75,122 @@ HTMLAttachmentElement::~HTMLAttachmentElement() = default;
     return createRenderer<RenderAttachment>(*this, WTFMove(style));
 }
 
+void HTMLAttachmentElement::invalidateRender()
```
`invalidateRenderer`

squelart replied:
I thought "render" was also a noun but it's only a verb; I meant "rendering".
I didn't want to imply that I'm invalidating the **renderer** object, but the rendering (that the renderer would have produced) instead.
So, okay with `invalidateRendering()`? But if you really prefer `invalidateRenderer()` (or something else) please let me know.

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.h:
```diff
@@ -101,12 +110,19 @@ class HTMLAttachmentElement final : public HTMLElement {
     bool childShouldCreateRenderer(const Node&) const final;
 #endif
 
+    enum class Implementation { Legacy, Modern, LegacyUnderModern };
```
`enum class Implementation : uint8_t`

pxlcoder replied:
`LegacyUnderModern` is a confusing name – maybe `IconOnly`?

squelart replied:
"LegacyUnderModern" because it's really a legacy attachment that is always a child of, and is controlled by, a modern attachment.
It could be more than an icon: Thumbnail, placeholder...
Other suggestions?
(Note that eventually the whole legacy code should be removed, once we've fully transitioned.)

pxlcoder commented on Source/WebCore/rendering/AttachmentLayout.mm:
```diff
@@ -135,7 +135,7 @@ static Color titleTextColorForAttachment(const RenderAttachment& attachment, Att
 void AttachmentLayout::layOutSubtitle(const RenderAttachment& attachment)
 {
-    auto& subtitleText = attachment.attachmentElement().attributeWithoutSynchronization(HTMLNames::subtitleAttr);
+    String subtitleText = attachment.attachmentElement().attachmentSubtitleForDisplay();
```
Can leave as `auto&`.

squelart replied:
Nope, can't do in this case: `attributeWithoutSynchronization` used to return a `const AtomString&`, but now `attachmentSubtitleForDisplay` returns a `String` temporary object, which cannot be bound to a non-const lvalue reference.
`const auto&` would work, if you preferred?

pxlcoder replied:
Looks like `String` is used in other call sites, so this is fine.

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -70,14 +75,122 @@ HTMLAttachmentElement::~HTMLAttachmentElement() = default;
+        ASSERT(!attachment->renderer()); // Switch to modern style *must* be done before renderer is created!
+        attachment->m_implementation = Implementation::Modern;
+        if (auto shadow = attachment->userAgentShadowRoot())
+            attachment->ensureModernShadowTree(*shadow);
```
Does this code ever run? I'm not sure how the element would already have a `userAgentShadowRoot` at creation time.

nt1m replied:
```suggestion
        if (auto shadowRoot = attachment->userAgentShadowRoot())
            attachment->ensureModernShadowTree(*shadowRoot);
```

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -70,14 +75,122 @@ HTMLAttachmentElement::~HTMLAttachmentElement() = default;
 {
-    return adoptRef(*new HTMLAttachmentElement(tagName, document));
+    Ref<HTMLAttachmentElement> attachment = adoptRef(*new HTMLAttachmentElement(tagName, document));
+    if (DeprecatedGlobalSettings::attachmentWideLayoutEnabled() && DeprecatedGlobalSettings::attachmentElementEnabled()) {
```
```suggestion
    if (document.settings().attachmentWideLayoutEnabled() && DeprecatedGlobalSettings::attachmentElementEnabled()) {
```

To avoid adding a new DeprecatedGlobalSetting.

squelart replied:
I used a DeprecatedGlobalSetting to be consistent with the existing `attachmentElementEnabled` and keep them together. But I'll have a look...

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -70,14 +75,122 @@ HTMLAttachmentElement::~HTMLAttachmentElement() = default;
 Ref<HTMLAttachmentElement> HTMLAttachmentElement::create(const QualifiedName& tagName, Document& document)
 {
-    return adoptRef(*new HTMLAttachmentElement(tagName, document));
+    Ref<HTMLAttachmentElement> attachment = adoptRef(*new HTMLAttachmentElement(tagName, document));
```
```suggestion
    Ref attachment = adoptRef(*new HTMLAttachmentElement(tagName, document));
```

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -70,14 +75,122 @@ HTMLAttachmentElement::~HTMLAttachmentElement() = default;
+    max-width: 500px;
+    background-color: lightgray;
+    border-radius: 8px;
+    font-family: "Lucida Grande";
```
You probably want a system font value: https://developer.mozilla.org/en-US/docs/Web/CSS/font#system_font_values

Although if used in the relevant contexts, the correct font should just inherit down.

squelart replied:
It will be used inside an editor with user-selected fonts around, so I think I need to specify our choice of font here.
Thanks for the tip about system fonts.

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -70,14 +75,122 @@ HTMLAttachmentElement::~HTMLAttachmentElement() = default;
     return createRenderer<RenderAttachment>(*this, WTFMove(style));
 }
 
+void HTMLAttachmentElement::invalidateRender()
```
```suggestion
void HTMLAttachmentElement::invalidateRenderer()
```

squelart replied:
I actually meant `invalidateRendering` (I'm not invalidating the renderer object itself). Okay with you?

nt1m commented on Source/WebCore/html/shadow/ShadowPseudoIds.h:
```diff
@@ -43,6 +43,12 @@ const AtomString& webkitCreditCardAutoFillButton();
+const AtomString& webkitAttachmentInfo();
+const AtomString& webkitAttachmentPreview();
+const AtomString& webkitAttachmentSubtitle();
+const AtomString& webkitAttachmentTitle();
```
can we guard them behind `ENABLE(ATTACHMENT_ELEMENT)`? We should absolutely not expose these to the web.

squelart replied:
I'm moving away from pseudo-ids, so these will be removed.

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -70,14 +75,122 @@ HTMLAttachmentElement::~HTMLAttachmentElement() = default;
+    container->setPseudo(ShadowPseudoIds::webkitAttachmentContainer());
+    root.appendChild(container);
+
+    m_innerLegacyAttachment = adoptRef(*new HTMLAttachmentElement(HTMLNames::attachmentTag, document()));
```
I don't feel too great about nested attachment elements as a long term approach. Do you have a plan to address this / alternatives considered? I know there was some discussion about using blob URLs for the icon.

If we are going to continue with the icon as a distinct element, this should be it's own internal element type (similar to `SpinButtonElement`)

squelart replied:
There are a few reasons I went this way:
- Using blob URLs was not easy (to me at least 😅), and also it felt it would have been quite expensive in both CPU and memory to perform multiple counter-acting conversions.
- We can keep the legacy code around while transitioning, so even a new WebKit library could still work with an older Mail app.
- The legacy attachment code does more than just display an image, it also handles requesting that icon/thumbnail when first rendering, showing an optional progress bar, and other operations, which would make it much more difficult to port to our "modern" element all at once.
- And it's not meant to stay long term, as eventually the legacy code should be removed once we only need the modern style and everything works with it as needed.

pxlcoder replied:
Sounds reasonable to me. Can you add a FIXME + bug, so that we consider improving this in the longer term.

nt1m commented on Source/WebCore/rendering/RenderAttachment.h:
```diff
@@ -66,9 +65,9 @@ class RenderAttachment final : public RenderReplaced {
 {
-    return downcast<RenderAttachment>(HTMLElement::renderer());
+    return dynamicDowncast<RenderAttachment>(HTMLElement::renderer());
 }
```
We should really delete this method, and downcast + assert in the right places. It's misleading since attachmentRenderer isn't always valid.

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.h:
```diff
@@ -101,12 +110,19 @@ class HTMLAttachmentElement final : public HTMLElement {
     bool childShouldCreateRenderer(const Node&) const final;
 #endif
 
+    enum class Implementation { Legacy, Modern, LegacyUnderModern };
```
Not sure what LegacyUnderModern means.

squelart replied:
A legacy attachment (drawing the old way) that lives in the shadow tree under a modern attachment. Other suggestions?

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -71,14 +76,80 @@ HTMLAttachmentElement::~HTMLAttachmentElement() = default;
 {
-    return adoptRef(*new HTMLAttachmentElement(tagName, document));
+    Ref attachment = adoptRef(*new HTMLAttachmentElement(tagName, document));
+    if (document.settings().attachmentWideLayoutEnabled() && DeprecatedGlobalSettings::attachmentElementEnabled()) {
```
The check for `DeprecatedGlobalSettings::attachmentElementEnabled()` is unnecessary, since you already can't create `HTMLAttachmentElement`s if it's false.

(see `HTMLAttachmentElement.idl`)

squelart replied:
Oops, good point! I should rephrase the setting description then, from "wide-layout implies attachment element" to "... requires ...".

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -30,17 +30,22 @@
 
 #include "AttachmentElementClient.h"
 #include "DOMURL.h"
+#include "DeprecatedGlobalSettings.h"
```
This will be an unnecessary include, following the comment below.

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -71,14 +76,80 @@ HTMLAttachmentElement::~HTMLAttachmentElement() = default;
+    root.appendChild(WTFMove(style));
+
+    auto container = HTMLDivElement::create(document());
+    static MainThreadNeverDestroyed<const AtomString> containerId("attachment-container"_s);
```
It might be nice to put these ids in their own method, similar to what's done in `ImageControlsMac`.

squelart replied:
It seemed heavy, since I only used the ids once. In `ImageControlsMac` they're used twice.
But it could be useful in the future, and it's more consistent between similar files, so I'll make the change.

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.h:
```diff
@@ -69,23 +70,30 @@ class HTMLAttachmentElement final : public HTMLElement {
     void requestIconWithSize(const FloatSize&) const;
     FloatSize iconSize() const { return m_iconSize; }
-    RenderAttachment* renderer() const;
+    void invalidaterendering();
```
Camel case.

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.h:
```diff
@@ -101,12 +109,19 @@ class HTMLAttachmentElement final : public HTMLElement {
     bool childShouldCreateRenderer(const Node&) const final;
 #endif
 
+    enum class Implementation: uint8_t { Legacy, Modern, LegacyUnderModern };
```
Still not a fan of the `LegacyUnderModern` name.

You mentioned:

> It could be more than an icon: Thumbnail, placeholder...

Unless I'm missing something, these are just icons right? A thumbnail icon, or a placeholder icon?

squelart replied:
Oh, for me, a "thumbnail" is an image that is a small representation of the **contents** of a document (like a minimized screenshot, or an extract), while an "icon" is also an image but an abstract representation of a **type** of documents (like the same PDF logo for all pdf files). Hence my confusion and resistance when you suggested "IconOnly".
Wikipedia seems to agree with me, however I've found another definition of thumbnail saying "an image of a graphic or document that is reduced in size so that it serves as an icon of a larger image", so I'm not so sure anymore.
In any case it shows that it could be confusing to *some* people! 😉

So how about we call it "ImageOnly"?
It should be less prone to confusion, and will better match what we're doing in the old rendering code (tests that skip rendering anything else than images/icons/thumbnails/placeholders).

pxlcoder replied:
Sounds good to me!

pxlcoder commented on Source/WebCore/html/shadow/attachmentElementShadow.css:
```diff
@@ -0,0 +1,57 @@
+div#attachment-container {
+    display: grid;
+    max-width: 500px;
+    background-color: lightgray; /* FIXME: Use semantic colors */
```
This FIXME should have an accompanying radar/bugzilla.

pxlcoder commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -1210,7 +1210,7 @@ bool DragController::startDrag(Frame& src, const DragState& state, OptionSet<Dra
     if (is<HTMLAttachmentElement>(element) && m_dragSourceAction.contains(DragSourceAction::Attachment)) {
         auto& attachment = downcast<HTMLAttachmentElement>(element);
-        auto* attachmentRenderer = attachment.renderer();
+        auto* attachmentRenderer = dynamicDowncast<RenderAttachment>(attachment.renderer());
```
Since we're changing this code, would you mind moving this down closer to where it's used?

(Underneath the `if (!dragImage)`)

pxlcoder commented on Source/WebCore/rendering/RenderThemeMac.mm:
```diff
@@ -1581,14 +1582,18 @@ static void paintAttachmentPlaceholderBorder(const RenderAttachment& attachment,
     bool usePlaceholder = validProgress && !progress;
 
-    paintAttachmentIconBackground(attachment, context, layout);
+    if (!element.isLegacyUnderModern())
```
Re-naming `LegacyUnderModern` would also make this code cleaner. This method should really describe the behavior (along the lines of "drawsIconOnly", rather than discuss the "type" of element).

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.h:
```diff
@@ -69,23 +70,30 @@ class HTMLAttachmentElement final : public HTMLElement {
     void setImageMenuEnabled(bool value) { m_isImageMenuEnabled = value; }
 #endif
 
+    bool isModern() const { return m_implementation == Implementation::Modern; }
```
Nit: This method is only used in one place, so maybe you don't need it? If you do keep it, it should be `private`.

pxlcoder reviewed (approved):
I think @hortont424 or @whsieh should take a quick look too 😄

whsieh commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -71,14 +75,103 @@ HTMLAttachmentElement::~HTMLAttachmentElement() = default;
+    m_innerLegacyAttachment->m_thumbnail = WTFMove(m_thumbnail);
+    m_innerLegacyAttachment->m_icon = WTFMove(m_icon);
+    m_innerLegacyAttachment->m_iconSize = m_iconSize;
+    m_innerLegacyAttachment->setIdAttribute(attachmentPreviewIdentifier());
```
Just wondering — does this mean that we'd have two `API::Attachment` objects in the UI process, one corresponding the outer "modern" attachment and another corresponding to the inner legacy attachment?

squelart replied:
Like the rest of the shadow DOM elements, the inner attachment element is an implementation detail of how to display icons/thumbnails inside the outer (public) attachment, so it shouldn't be visible anywhere; at least the apps using attachments shouldn't need to know about it.
However I see that *all* attachments call `Document::didInsertAttachmentElement`, and through a series of calls&messages, end in `WebPageProxy::didInsertAttachmentWithIdentifier` where a corresponding `API::Attachment` is created.
I admit I'm not sure what this means. If the UIProcess' `API::Attachment` is needed to display it, then that should be fine and work as expected. Otherwise we could theoretically drop the initial `Document::didInsertAttachmentElement` from inner legacy attachments. I shall explore this further, and I may need to talk to someone knowledgeable to resolve this soon!

nt1m replied:
If you can find a way to extract the code that draws the file icon (without using a nested attachment element), that would be really useful. `<input type="file">` could also use such code to display file icons if we rewrite it to use shadow DOM.

squelart replied:
Sorry, I should have been clearer: The inner attachment does more than just draw images, it actually requests the proper icon/thumbnail for a given file, it also draws the progress bar; and some of this is different between platforms.
All this should eventually be moved to the modern element layout&rendering, but it was more work&time than using this nesting trick for now -- at least for me!

squelart committed:
Opt-in wide-layout attachment style using Shadow DOM tree
https://bugs.webkit.org/show_bug.cgi?id=251830
rdar://105103464

Reviewed by Aditya Keerthi.

Note that the modern attachment element uses an inner legacy attachment just to display the correct thumbnail or icon.

* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::HTMLAttachmentElement):
(WebCore::HTMLAttachmentElement::ensureModernShadowRoot):
(WebCore::HTMLAttachmentElement::didAddUserAgentShadowRoot):
(WebCore::HTMLAttachmentElement::ensureModernShadowTree):
(WebCore::HTMLAttachmentElement::createElementRenderer): Deleted.
(WebCore::HTMLAttachmentElement::getAttachmentIdentifier): Deleted.
(WebCore::HTMLAttachmentElement::copyNonAttributePropertiesFromElement): Deleted.
(WebCore::HTMLAttachmentElement::archiveResourceURL): Deleted.
(WebCore::HTMLAttachmentElement::file const): Deleted.
(WebCore::HTMLAttachmentElement::blobURL const): Deleted.
(WebCore::HTMLAttachmentElement::setFile): Deleted.
(WebCore::HTMLAttachmentElement::insertedIntoAncestor): Deleted.
(WebCore::HTMLAttachmentElement::removedFromAncestor): Deleted.
(WebCore::HTMLAttachmentElement::ensureUniqueIdentifier): Deleted.
(WebCore::HTMLAttachmentElement::setUniqueIdentifier): Deleted.
(WebCore::HTMLAttachmentElement::enclosingImageElement const): Deleted.
(WebCore::HTMLAttachmentElement::parseAttribute): Deleted.
(WebCore::HTMLAttachmentElement::attachmentTitle const): Deleted.
(WebCore::HTMLAttachmentElement::attachmentTitleForDisplay const): Deleted.
(WebCore::HTMLAttachmentElement::attachmentType const): Deleted.
(WebCore::HTMLAttachmentElement::attachmentPath const): Deleted.
(WebCore::HTMLAttachmentElement::updateAttributes): Deleted.
(WebCore::mimeTypeIsSuitableForInlineImageAttachment): Deleted.
(WebCore::HTMLAttachmentElement::updateEnclosingImageWithData): Deleted.
(WebCore::HTMLAttachmentElement::updateThumbnail): Deleted.
(WebCore::HTMLAttachmentElement::updateIcon): Deleted.
(WebCore::HTMLAttachmentElement::requestIconWithSize const): Deleted.
(WebCore::HTMLAttachmentElement::childShouldCreateRenderer const): Deleted.
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/rendering/AttachmentLayout.mm:
(WebCore::AttachmentLayout::layOutSubtitle):
(WebCore::AttachmentLayout::AttachmentLayout):

Canonical link: https://commits.webkit.org/260296@main
